File: VinCommon/src/com/vinecom/common/tool/topK/ArrayTopK.java
package com.vinecom.common.tool.topK;

import com.vinecom.common.data.Tuple2;

import java.util.Arrays;

/**
 * Created by ndn on 7/23/2015.
 */
public class ArrayTopK<A> implements TopK<A> {
    private Tuple2<A, Double>[] arr;
    private int iMin = 0;
    private double vMin = Double.POSITIVE_INFINITY;
    private int size = 0;
    private int k;

    public ArrayTopK(int k) {
        this.arr = new Tuple2[k];
        this.k = k;
    }

    @Override
    public TopK<A> put(Tuple2<A, Double> v) {
        if (size < k) {
            arr[size] = v;
            if (v._2 < vMin) {
                vMin = v._2;
                iMin = size;
            }
            size += 1;
        }
        else if (v._2 > vMin) {
            arr[iMin] = v;
            iMin = 0;
            vMin = arr[0]._2;
            for (int i = 1; i < k; i++) {
                if (arr[i]._2 < vMin) {
                    iMin = i;
                    vMin = arr[i]._2;
                }
            }
        }

        return this;
    }

    @Override
    public TopK<A> put(A a, double d) {
        return put(new Tuple2<A, Double>(a, d));
    }

    @Override
    public TopK<A> put(TopK<A> other) {
        for (Tuple2<A, Double> e : other.all()) {
            put(e);
        }

        return this;
    }

    @Override
    public int size() {
        return this.size;
    }

    @Override
    public Tuple2<A, Double>[] real() {
        Tuple2<A, Double>[] copy = all();
        Arrays.sort(copy);
        return copy;
    }

    @Override
    public Tuple2<A, Double>[] all() {
        return Arrays.copyOf(this.arr, this.size);
    }
}

File: REC/src/main/scala/com/vinecom/recommend/tools/topK/ArrayTopK.scala
package com.vinecom.recommend.tools.topK

/**
 * Lớp chỉ chứa k phần tử lớn nhất của tập các phần tử được lần lượt đưa vào
 * Created by ndn on 3/17/2015.
 */
class ArrayTopK[A](k: Int) extends TopK[A]{
  val arr = new Array[(A, Double)](k)
  private val size1 = k - 1
  private var iMin = 0
  private var vMin = Double.PositiveInfinity
  var size = 0

  override def put(v: (A, Double)): TopK[A] = {
    if (size < k) {
      arr(size) = v
      if (v._2 < vMin) {
        vMin = v._2
        iMin = size
      }
      size += 1
    }
    else if (v._2 > vMin) {
      arr(iMin) = v
      iMin = 0
      vMin = arr(0)._2
      for (i <- 1 to size1)
        if (arr(i)._2 < vMin) {
          iMin = i
          vMin = arr(i)._2
        }
    }
    this
  }

  override def put(other: TopK[A]): TopK[A] = {
    val size1 = other.size - 1
    for (i <- 0 to size1) {
      put(other.get(i))
    }
    this
  }

  /**
   * @return Trả về mảng với kích thước thật và được sắp xếp theo thứ tự điểm giảm dần
   */
  override def real = arr.take(size).sortBy(_._2)(Ordering[Double].reverse)

  override def get(index: Int): (A, Double) = arr(index)
}
